lib/duo.3
.\"
.\" SPDX-License-Identifier: GPL-2.0-with-classpath-exception
.\" Copyright (c) 2023 Cisco Systems, Inc. and/or its affiliates
.\" All rights reserved.
.\"
.Dd October 31, 2010
.Dt DUO 3
.Os
.Sh NAME
.Nm duo
.Nd Duo authentication service
.Sh SYNOPSIS
.Fd #include <duo.h>
.Ft duo_t *
.Fn duo_open "const char *ikey" "const char *skey" "const char *progname" "const char *cafile"
.Ft void
.Fn duo_set_conv_funcs "duo_t *d" "char *(*conv_prompt)(void *conv_arg, const char *, char *, size_t)" "void (*conv_status)(void *conv_arg, const char *msg)" "void *conv_arg"
.Ft void
.Fn duo_set_host "duo_t *d" "const char *hostname"
.Ft void
.Fn duo_set_ssl_verify "duo_t *d" "int bool"
.Ft duo_code_t
.Fn duo_login "duo_t *d" "const char *username" "const char *client_ip" "int flags" "const char *command"
.Ft const char *
.Fn duo_geterr "duo_t *d"
.Ft void
.Fn duo_close "duo_t *d"
.Sh DESCRIPTION
The
.Nm
API provides access to the Duo two-factor authentication service.
.Pp
.Fn duo_open
is used to obtain a handle to the Duo service.
.Fa ikey
and
.Fa skey
are the required integration and secret keys, respectively, for a Duo customer
account.
.Fa progname
identifies the program to the Duo service.
.Fa cafile
should be 
.Li NULL
or the pathname of a PEM-format CA certificate to override the default.
.Pp
.Fn duo_set_conv_funcs
may be used to override the internal user conversation functions.
.Fa conv_prompt
is called to present the user a login menu and 
.Fa prompt ,
and gather their response, returning 
.Fa buf 
or NULL on error. It may be set to NULL if automatic login is
specified with DUO_FLAG_AUTO.
.Fa conv_status
is called to display status messages to the user, and may be NULL if
no status display is needed.
.Fa conv_arg
is passed as the first argument to these conversation functions.
.Pp
.Fn duo_set_host
may be used to override the default Duo API host.
.Pp
.Fn duo_set_ssl_verify
may be used to override SSL certificate verification (enabled by
default).
.Pp
.Fn duo_login
performs secondary authentication via the Duo service for the specified
.Fa username Ns .
.Fa client_ip
is the source IP address of the connection to be authenticated, or
.Li NULL
to specify the local host. The following bitmask values are defined for
.Fa flags :
.Pp
.Bl -tag -width "DUO_SERVER_ERROR" -compact -offset indent
.It Li DUO_FLAG_AUTO
Attempt authentication without prompting the user, using their default
out-of-band authentication factor.
.It Li DUO_FLAG_SYNC
Do not report incremental status during authentication (e.g. voice
callback progress) - only issue one status message per authentication
attempt.
.El
.Pp
If not
.Li NULL ,
the 
.Fa command
to be authorized will be displayed during push authentication.
.Pp
.Fn duo_geterr
returns a description of the last-seen error on the specified Duo API
handle. The returned constant string should not be modified or freed
by the caller.
.Pp
.Fn duo_close
closes and frees the specified Duo API handle.
.Sh RETURN VALUES
.Fn duo_open
returns a pointer to the configured Duo API handle, or 
.Li NULL
on failure.
.Pp
.Fn duo_login
returns status codes of type 
.Ft duo_code_t ,
which may have the following values:
.Pp
.Bl -tag -width "DUO_SERVER_ERROR" -compact -offset indent
.It Li DUO_OK
User authenticated
.It Li DUO_FAIL
User failed to authenticate
.It Li DUO_ABORT
User denied by policy
.It Li DUO_LIB_ERROR
Unexpected library error
.It Li DUO_CONN_ERROR
Duo service unreachable
.It Li DUO_CLIENT_ERROR
Invalid client parameters to API call
.It Li DUO_SERVER_ERROR
Duo service error
.El
.Pp
In the event of a DUO_*_ERROR return, 
.Xr duo_geterr
may be called to recover a human-readable error message.
.Pp
.Fn duo_geterr
returns a constant string which should not be modified or freed by the
caller.
.Sh SEE ALSO
.Xr pam_duo 8 ,
.Xr login_duo 1
.Sh AUTHORS
Duo Security
.Aq support@duosecurity.com


CHANGES
duo_unix-2.0.1:
- The support script collects a few additional files for troubleshooting
- Duo API calls now use SHA512 instead of SHA1 as the HMAC algorithm

duo_unix-2.0.0:
- Changed the behavior of `su` when the target user is not root.  The target user will need to complete 2FA rather than the original user.
- login_duo resets the SIGPIPE handler when it closes its connection.
- Added logging when Duo is invoked, to assist troubleshooting.
- Updated package signing to SHA512

duo_unix-1.12.1:
- Updated Unity to 2.5.2
- Added support for Fedora 34
- Removed support for Centos 8
- Added support for Centos Stream 8
- Added support for Centos Stream 9
- Added support for Ubuntu 22.04

duo_unix-1.12.0:
- Switched from BSON to JSON as a data interchange format
- Switched from Cram to python `unittest` for testing

duo_unix-1.11.5:
- Added support for Debian 11
- Removed support for Debian 8
- Removed support for CentOS 6
- Fixed MOTD display for non-interactive sessions
- The support tool now also collects the sudo PAM configuration file
- Updated pinned certificates

duo_unix-1.11.4:
- Added support for Ubuntu 20.04
- Added support tool to collect information (e.g. logs and PAM stacks) for debugging purposes

duo_unix-1.11.3:
- Added support for RedHat 8, CentOS 8, and Debian 10
- Improved validation of BSON messages

duo_unix-1.11.2:
- Added recommended Kerberos configuration for Duo Unix to our documentation, found at https://help.duo.com/s/article/5085. Thanks to Neal Poole at Facebook for bringing expertise and attention to this topic.
- Updated SELinux policy to allow local logins to use the pam_duo PAM module and made sshd configurable
- Added support for spaces in group names when escaped with backslashes in pam_duo.conf and login_duo.conf
- Test infrastructure updates

duo_unix-1.11.1:
- Fixed bug causing console login to fail on certain systems

duo_unix-1.11.0:

- Added support for GECOS field parsing based on user-supplied delimiter
- Updated README to include development/testing steps
- Minor test infrastructure updates

duo_unix-1.10.5:

- Fixed an accidental null pointer free on systems where getaddrinfo() is unsuccessful

duo_unix-1.10.4:

- Removed failmode decision from auth endpoint and moved it to only preauth according to standards in our other integrations
- Updated Duo Unix to speak up to TLS 1.2
- Support for LibreSSL 2.7.0 and up
- Minor memory leak fixes
- Output message when user is locked out

duo_unix-1.10.3:

- Added support for http_proxy with SELinux enabled

duo_unix-1.10.2:

- Added default failmode values in config files

duo_unix-1.10.1:

- Fixed bug causing automated tests to fail on OSX
- Addressed an issue which kept configuration secrets in memory for longer than necessary

duo_unix-1.10.0:

- Added LibreSSL support
- Added additional GECOS parsing support
- Increased OSX group count

duo_unix-1.9.21:

- PSA-2017-002: Only allow http_proxy to be defined in configuration file instead of environment

duo_unix-1.9.20:

- Fix installation on AIX systems
- Add support for using OpenSSL 1.1.0
- Link libduo statically to address issues with the ldconfig cache and incompatibilities between versions
- Fixed a bug that produced incorrect SNI when using a proxy

duo_unix-1.9.19:

- Restore the http_proxy environment variable after Duo is done
- Added https_timeout config option to pam_duo
- Handles missing shell and adds default if not specified in getpwuid
- Add SNI support and a guard for systems that don't support SNI
- Bug fixes for timeouts and fallback ip addresses

duo_unix-1.9.18:

- Added HTTP proxy connection error handling
- Improved compatibility with Solaris and AIX

duo_unix-1.9.17:

- Fixed PAM return code issue

duo_unix-1.9.16:

- Test fixes
- Compilation fixes

duo_unix-1.9.15:

- SELinux policy module package support
- PAM module improvements
- Removed deprecated SHA1 Entrust CA

duo_unix-1.9.14:

- Added SELinux policy module
- Improve poll(2) error handling

duo_unix-1.9.13:

- Bugfixes for signal handling

duo_unix-1.9.12:

- Include https_timeout configuration parameter
- IPv6 support on systems that have getaddrinfo

duo_unix-1.9.11:

- Improve compatibility with FreeBSD 10.

duo_unix-1.9.10:

- Use the correct timeout when polling.

duo_unix-1.9.9:

- Use poll(2) instead of select(2) for timeouts to support busy
  systems with many open file descriptors.
- Send User-Agent header with each request.

duo_unix-1.9.8:

- Improve support for SHA2 in HTTPS.

duo_unix-1.9.7:

- Allow using accept_env_factor with SSH.
- Allow using autopush with PAM on Mac OS X.

duo_unix-1.9.6:

- Update HTTPS CA certificates.

duo_unix-1.9.5:

- Fix issues running 'make check'

- Remove accept_env_factor from pam_duo manpage, as it will not work

duo_unix-1.9.4:

- Send codes / push requests using $DUO_PASSCODE environment variable

- Fix error in 1.9.3 changelog :)

- pam_duo is feature-par with login_duo (autopush, prompts)

- Internal refactoring


lib/duo.c
/*
 * SPDX-License-Identifier: GPL-2.0-with-classpath-exception
 *
 * duo.c
 *
 * Copyright (c) 2023 Cisco Systems, Inc. and/or its affiliates
 * All rights reserved.
 */

#include "config.h"

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>

#include <arpa/inet.h>
#include <netinet/in.h>

#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <stdarg.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <openssl/bio.h>
#include <openssl/evp.h>
#include <openssl/hmac.h>
#include <openssl/sha.h>
#include <openssl/ssl.h>

#include "util.h"
#include "duo.h"
#include "parson.h"
#include "duo_private.h"
#include "ini.h"
#include "urlenc.h"

#define DUO_LIB_VERSION     "libduo/" PACKAGE_VERSION
#define DUO_API_VERSION     "/rest/v1"
#define AUTOPUSH_MSG        "Autopushing login request to phone..."
#define AUTOPHONE_MSG       "Calling your phone..."
#define AUTODEFAULT_MSG     "Using default second-factor authentication."
#define ENV_VAR_MSG         "Reading $DUO_PASSCODE..."

/*
 * Finding the maximum length for the machine's hostname
 * Idea and technique originated from https://github.com/openssh/openssh-portable
 */
#ifndef HOST_NAME_MAX
# include "netdb.h" /* for MAXHOSTNAMELEN */
# if defined(_POSIX_HOST_NAME_MAX)
#  define HOST_NAME_MAX _POSIX_HOST_NAME_MAX
# elif defined(MAXHOSTNAMELEN)
#  define HOST_NAME_MAX MAXHOSTNAMELEN
# else
#  define HOST_NAME_MAX 255
# endif
#endif /* HOST_NAME_MAX */

/* For sizing buffers; sufficient to cover the longest possible DNS FQDN */
#define DNS_MAXNAMELEN 256

static char *
__prompt_fn(void *arg, const char *prompt, char *buf, size_t bufsz)
{
    printf("%s", prompt);
    fflush(stdout);
    return (fgets(buf, bufsz, stdin));
}

static void
__status_fn(void *arg, const char *msg)
{
    printf("%s\n", msg);
}

struct duo_ctx *
duo_open(const char *host, const char *ikey, const char *skey,
    const char *progname, const char *cafile, int https_timeout, const char* http_proxy)
{
    struct duo_ctx *ctx;

    if ((ctx = calloc(1, sizeof(*ctx))) == NULL ||
            (ctx->host = strdup(host)) == NULL ||
            (ctx->ikey = strdup(ikey)) == NULL ||
            (ctx->skey = strdup(skey)) == NULL) {
        return (duo_close(ctx));
    }
    if (asprintf(&ctx->useragent, "%s (%s) libduo/%s",
            progname, CANONICAL_HOST, PACKAGE_VERSION) == -1) {
        return (duo_close(ctx));
    }
    if (https_init(cafile, http_proxy) != HTTPS_OK) {
        ctx = duo_close(ctx);
    } else {
        ctx->conv_prompt = __prompt_fn;
        ctx->conv_status = __status_fn;
        ctx->https_timeout = https_timeout;
    }

    return (ctx);
}

int
duo_parse_config(const char *filename,
    int (*callback)(void *arg, const char *section,
    const char *name, const char *val), void *arg)
{
    FILE *fp;
    struct stat st;
    int fd, ret;

    if ((fd = open(filename, O_RDONLY)) < 0) {
        return (-1);
    }
    if (fstat(fd, &st) < 0 || (fp = fdopen(fd, "r")) == NULL) {
        close(fd);
        return (-1);
    }
// Windows permissions don't work like unix permissions, so lets not
// try this like this - we'll trust the user to get it right for now...
// eventually this should be replaced with a path that actually does this right
#ifndef WIN32
    if ((st.st_mode & (S_IRGRP|S_IROTH)) != 0) {
        fclose(fp);
        return (-2);
    }
#endif
    ret = ini_parse(fp, callback, arg);
    fclose(fp);
    return (ret);
}

static duo_code_t
duo_reset(struct duo_ctx *ctx)
{
    int i;

    for (i = 0; i < ctx->argc; i++) {
        free(ctx->argv[i]);
        ctx->argv[i] = NULL;
    }
    ctx->argc = 0;
    *ctx->err = '\0';

    return (DUO_OK);
}

struct duo_ctx *
duo_close(struct duo_ctx *ctx)
{
    if (ctx != NULL) {
        if (ctx->https != NULL) {
            https_close(&ctx->https);
        }
        duo_reset(ctx);
        free(ctx->host);

        // We need to add 1 here for the terminating \0 byte which strlen doesn't include
        if (ctx->ikey != NULL) {
            duo_zero_free(ctx->ikey, strlen(ctx->ikey) + 1);
            ctx->ikey = NULL;
        }
        if (ctx->skey != NULL) {
            duo_zero_free(ctx->skey, strlen(ctx->skey) + 1);
            ctx->skey = NULL;
        }
        if (ctx->useragent != NULL) {
            duo_zero_free(ctx->useragent, strlen(ctx->useragent) + 1);
            ctx->useragent = NULL;
        }

        free(ctx);
    }
    return (NULL);
}

void
duo_set_conv_funcs(struct duo_ctx *ctx,
    char *(*prompt_fn)(void *arg, const char *prompt, char *buf, size_t bufsz),
    void (*status_fn)(void *arg, const char *msg),
    void *arg)
{
    ctx->conv_prompt = prompt_fn;
    ctx->conv_status = status_fn;
    ctx->conv_arg = arg;
}

void
duo_reset_conv_funcs(struct duo_ctx *ctx)
{
    ctx->conv_prompt = __prompt_fn;
    ctx->conv_status = __status_fn;
}

duo_code_t
duo_add_param(struct duo_ctx *ctx, const char *name, const char *value)
{
    duo_code_t ret;
    char *k, *v, *p;

    if (name == NULL || value == NULL || strlen(name) == 0 || strlen(value) == 0) {
        return (DUO_CLIENT_ERROR);
    }
    ret = DUO_LIB_ERROR;

    k = urlenc_encode(name);
    v = urlenc_encode(value);

    if (k && v && ctx->argc + 1 < (sizeof(ctx->argv) / sizeof(ctx->argv[0]))
            && (asprintf(&p, "%s=%s", k, v) > 2)) {
        ctx->argv[ctx->argc++] = p;
        ret = DUO_OK;
    }

    free(k);
    free(v);

    return (ret);
}

duo_code_t
duo_add_optional_param(struct duo_ctx *ctx, const char *name, const char *value)
{
    /* Wrapper around duo_add_param for optional arguments.
       If a parameter's value doesn't exist we don't add the param.
    */
    if (value == NULL || strlen(value) == 0) {
        return DUO_OK;
    }
    else {
        return duo_add_param(ctx, name, value);
    }
}

static void
_duo_seterr(struct duo_ctx *ctx, const char *fmt, ...)
{
    va_list ap;

    va_start(ap, fmt);
    vsnprintf(ctx->err, sizeof(ctx->err), fmt, ap);
    va_end(ap);
}


static void
_duo_get_hostname(char *dns_fqdn, size_t dns_fqdn_size)
{
    struct addrinfo hints, *info, *p;
    char hostname[HOST_NAME_MAX + 1];

    /* gethostname may not insert a null terminator when it needs to truncate the hostname.
     * See gethostname's man page under "Description" for more info.
     */
    hostname[HOST_NAME_MAX] = '\0';
    gethostname(hostname, HOST_NAME_MAX);
    memset(&hints, 0, sizeof(hints));

    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_flags = AI_CANONNAME;
    strlcpy(dns_fqdn, hostname, dns_fqdn_size);

    if (getaddrinfo(hostname, NULL, &hints, &info) == 0) {
        if(info->ai_canonname != NULL && strlen(info->ai_canonname) > 0) {
            strlcpy(dns_fqdn, info->ai_canonname, dns_fqdn_size);
        }
        freeaddrinfo(info);
    }
}

int
_duo_add_hostname_param(struct duo_ctx *ctx)
{
    char dns_fqdn[DNS_MAXNAMELEN];
    _duo_get_hostname(dns_fqdn, sizeof(dns_fqdn));

    return duo_add_optional_param(ctx, "hostname", dns_fqdn);
}

int _duo_add_failmode_param(struct duo_ctx *ctx, const int failmode)
{
    const char *failmode_str = (failmode == DUO_FAIL_SECURE) ? ("closed") : ("open");

    return duo_add_optional_param(ctx, "failmode", failmode_str);
}

#define _JSON_FIND_OBJECT(out_obj, in_obj, name, json_value) do { \
  out_obj = json_object_get_object(in_obj, name); \
  if (out_obj == NULL) { \
        _duo_seterr(ctx, "JSON missing valid '%s'", name);  \
        _JSON_VALUE_FREE(json_value); \
        return (DUO_SERVER_ERROR); \
  } \
} while(0)

#define _JSON_FIND_STRING(buf, json_obj, name, json_value) do { \
  buf = json_object_get_string(json_obj, name); \
  if (buf == NULL) { \
        _duo_seterr(ctx, "JSON missing valid '%s'", name);  \
        _JSON_VALUE_FREE(json_value); \
        return (DUO_SERVER_ERROR); \
  } \
} while(0)

# define _JSON_VALUE_FREE(value) do { \
  json_value_free(value); \
  value = NULL; \
} while(0)

static duo_code_t
_duo_json_response(struct duo_ctx *ctx) {
    JSON_Value *json;
    JSON_Object *json_obj;
    char *p;
    int code = DUO_SERVER_ERROR;

    json = json_parse_string(ctx->body);
    if(json == NULL) {
        _duo_seterr(ctx, "invalid JSON response");
        return (DUO_SERVER_ERROR);
    }
    json_obj = json_value_get_object(json);

    _JSON_FIND_STRING(p, json_obj, "stat", json);
    if (strcasecmp(p, "OK") == 0) {
        code = DUO_OK;
    }
    if (strcasecmp(p, "FAIL") == 0) {
        char *message;
        code = json_object_get_number(json_obj, "code");
        // json_object_get_number will return 0 if "code" not found
        if (code == 0) {
               _duo_seterr(ctx, "JSON missing valid 'code'");
               _JSON_VALUE_FREE(json);
               return (DUO_SERVER_ERROR);
        }
        _JSON_FIND_STRING(message, json_obj, "message", json);
        _duo_seterr(ctx, "%d: %s", code, message);
        code = DUO_FAIL;
    }
    _JSON_VALUE_FREE(json);
    return code;
}

static duo_code_t
duo_call(struct duo_ctx *ctx, const char *method, const char *uri, int msecs)
{
    int i, code, err, ret;

    code = 0;
    ctx->body = NULL;
    ctx->body_len = 0;

    for (i = 0; i < 3; i++) {
        if (ctx->https == NULL &&
            (err = https_open(&ctx->https, ctx->host, ctx->useragent)) != HTTPS_OK) {
            if (err == HTTPS_ERR_SERVER) {
                sleep(1 << i);
                continue;
            }
            break;
        }
        if ((err = https_send(ctx->https, method, uri,
                    ctx->argc, ctx->argv, ctx->ikey, ctx->skey, ctx->useragent)) == HTTPS_OK &&
            (err = https_recv(ctx->https, &code,
                &ctx->body, &ctx->body_len, msecs)) == HTTPS_OK) {
            break;
        }
        https_close(&ctx->https);
    }
    duo_reset(ctx);

    if (code == 0) {
        ret = DUO_CONN_ERROR;
        _duo_seterr(ctx, "Couldn't connect to %s: %s\n",
            ctx->host, https_geterr());
    } else if (code / 100 == 2) {
        /* 2xx indicates DUO_OK */
        ret = DUO_OK;
    } else if (code == 401) {
        /* 401 indicates an invalid ikey or skey */
        ret = DUO_CLIENT_ERROR;
        _duo_seterr(ctx, "Invalid ikey or skey");
    } else if (code / 100 == 5) {
        /* 5xx indicates an internal server error */
        ret = DUO_SERVER_ERROR;
        _duo_seterr(ctx, "HTTP %d", code);
    } else {
        /* abort on any other HTTP codes */
        ret = DUO_ABORT;
        _duo_seterr(ctx, "HTTP %d", code);
    }
    return (ret);
}

const char *
duo_geterr(struct duo_ctx *ctx)
{
    return (ctx->err[0] ? ctx->err : NULL);
}

duo_code_t
_duo_preauth(struct duo_ctx *ctx, const char *username,
    const char *client_ip, const int failmode)
{
    duo_code_t ret;
    const char *p;
    JSON_Value *json;
    JSON_Object *json_obj;

    /* Check preauth result */
    if (duo_add_param(ctx, "user", username) != DUO_OK) {
        return (DUO_LIB_ERROR);
    }

    if (duo_add_optional_param(ctx, "ipaddr", client_ip) != DUO_OK) {
        return (DUO_LIB_ERROR);
    }

    if(_duo_add_hostname_param(ctx) != DUO_OK) {
        return (DUO_LIB_ERROR);
    }

    if(_duo_add_failmode_param(ctx, failmode) != DUO_OK) {
        return (DUO_LIB_ERROR);
    }

    if ((ret = duo_call(ctx, "POST", DUO_API_VERSION "/preauth.json", ctx->https_timeout)) != DUO_OK ||
         (ret = _duo_json_response(ctx)) != DUO_OK)
    {
        return (ret);
    }

    json = json_parse_string(ctx->body);
    json_obj = json_value_get_object(json);
    JSON_Object *response;
    _JSON_FIND_OBJECT(response, json_obj, "response", json);
    _JSON_FIND_STRING(p, response, "result", json);
    if (p == NULL) {
        _duo_seterr(ctx, "JSON invalid 'result': %s", p);
        ret = DUO_SERVER_ERROR;
    } else if (strcasecmp(p, "auth") != 0) {
        char *output;
        _JSON_FIND_STRING(output, response, "status", json);
        if (strcasecmp(p, "allow") == 0) {
                        _duo_seterr(ctx, "%s", output);
            ret = DUO_OK;
        } else if (strcasecmp(p, "deny") == 0) {
            _duo_seterr(ctx, "%s", output);
            if (ctx->conv_status != NULL) {
                ctx->conv_status(ctx->conv_arg, output);
            }
            ret = DUO_ABORT;
        } else if (strcasecmp(p, "enroll") == 0) {
            if (ctx->conv_status != NULL) {
                ctx->conv_status(ctx->conv_arg, output);
            }
            _duo_seterr(ctx, "User enrollment required");
            ret = DUO_ABORT;
        } else {
            _duo_seterr(ctx, "JSON invalid 'result': %s", p);
            ret = DUO_SERVER_ERROR;
        }
    } else {
        ret = DUO_CONTINUE;
    }
    _JSON_VALUE_FREE(json);
    return (ret);
}

duo_code_t
_duo_prompt(struct duo_ctx *ctx, int flags, char *buf,
    size_t sz,  char *p, size_t sp)
{
    char *pos, *passcode;

    passcode = getenv(DUO_ENV_VAR_NAME);

    if ((flags & DUO_FLAG_ENV) && (passcode != NULL)) {
        if (strlcpy(p, passcode, sp) >= sp) {
            return (DUO_LIB_ERROR);
        }
        if (ctx->conv_status != NULL) {
            ctx->conv_status(ctx->conv_arg, ENV_VAR_MSG);
        }
        return (DUO_CONTINUE);
    } else if ((flags & DUO_FLAG_AUTO) != 0) {
        /* Find default OOB factor for automatic login */
        JSON_Value *json = json_parse_string(ctx->body);
        JSON_Object *json_obj = json_value_get_object(json);
        JSON_Object *response;
        JSON_Object *factors;
        _JSON_FIND_OBJECT(response, json_obj, "response", json);
        _JSON_FIND_OBJECT(factors, response, "factors", json);

        const char* default_factor;
        _JSON_FIND_STRING(default_factor, factors, "default", json);
        if (ctx->conv_status) {
            if ((pos = strstr(default_factor, "push"))) {
                ctx->conv_status(ctx->conv_arg, AUTOPUSH_MSG);
            } else if ((pos = strstr(default_factor, "phone"))) {
                ctx->conv_status(ctx->conv_arg, AUTOPHONE_MSG);
            } else {
                ctx->conv_status(ctx->conv_arg, AUTODEFAULT_MSG);
            }
        }
        if (strlcpy(p, default_factor, sp) >= sp) {
            _JSON_VALUE_FREE(json);
            return (DUO_LIB_ERROR);
        } else {
            _JSON_VALUE_FREE(json);
            return (DUO_CONTINUE);
        }
    } else {
        /* Prompt user for factor choice / token */
        if (ctx->conv_prompt == NULL) {
            _duo_seterr(ctx, "No prompt function set");
            return (DUO_CLIENT_ERROR);
        }
        JSON_Value *json = json_parse_string(ctx->body);
        JSON_Object *json_obj = json_value_get_object(json);
        JSON_Object *response;
        _JSON_FIND_OBJECT(response, json_obj, "response", json);

        const char* prompt;
        _JSON_FIND_STRING(prompt, response, "prompt", json);

        if (ctx->conv_prompt(ctx->conv_arg, prompt, buf, sz) == NULL) {
            _duo_seterr(ctx, "Error gathering user response");
            _JSON_VALUE_FREE(json);
            return (DUO_ABORT);
        }
        strtok(buf, "\r\n");

        JSON_Object *factors;
        _JSON_FIND_OBJECT(factors, response, "factors", json);

        // buf might not exist in factors JSON_Object, like if the user input
        // a passcode
        const char *factor_str = json_object_get_string(factors, buf);
        if (factor_str == NULL) {
            factor_str = buf;
        }
        if (strlcpy(p, factor_str, sp) >= sp) {
            _JSON_VALUE_FREE(json);
            return (DUO_LIB_ERROR);
        }
        _JSON_VALUE_FREE(json);
        return (DUO_CONTINUE);
    }
}

duo_code_t
duo_login(struct duo_ctx *ctx, const char *username,
    const char *client_ip, int flags, const char *command, const int failmode)
{
    duo_code_t ret;
    char buf[256];
    char *pushinfo = NULL;
    char p[256];
    int i;
    const char *local_ip;

    if (username == NULL) {
        _duo_seterr(ctx, "need username to authenticate");
        return (DUO_CLIENT_ERROR);
    }

    /* Check preauth status */
    if ((ret = _duo_preauth(ctx, username, client_ip, failmode)) != DUO_CONTINUE) {
        if(ret == DUO_SERVER_ERROR || ret == DUO_CONN_ERROR || ret == DUO_CLIENT_ERROR) {
            return (failmode == DUO_FAIL_SAFE) ? (DUO_FAIL_SAFE_ALLOW) : (DUO_FAIL_SECURE_DENY);
        }
        return (ret);
    }

    /* Handle factor selection */
    if ((ret = _duo_prompt(ctx, flags, buf, sizeof(buf), p, sizeof(p))) != DUO_CONTINUE) {
        return (ret);
    }

    /* Add request parameters */
    if (duo_add_param(ctx, "user", username) != DUO_OK ||
        duo_add_param(ctx, "factor", "auto") != DUO_OK ||
        duo_add_param(ctx, "auto", p) != DUO_OK ||
        duo_add_param(ctx, "async",
        (flags & DUO_FLAG_SYNC) ? "0" : "1") != DUO_OK) {
        return (DUO_LIB_ERROR);
    }

    /* Add client IP, if passed in */
    if (duo_add_optional_param(ctx, "ipaddr", client_ip) != DUO_OK) {
        return (DUO_LIB_ERROR);
    }

    if(_duo_add_hostname_param(ctx) != DUO_OK) {
        return (DUO_LIB_ERROR);
    }

    /* Add pushinfo parameters */
    local_ip = duo_local_ip();
    if (asprintf(&pushinfo, "Server+IP=%s&Command=%s",
        local_ip, command ? urlenc_encode(command) : "") < 0 ||
        duo_add_param(ctx, "pushinfo", pushinfo) != DUO_OK) {
        return (DUO_LIB_ERROR);
    }
    free(pushinfo);

    /* Try Duo authentication.  Only use the configured timeout if
     * the call is asynchronous, because async calls should return
     * immediately.
     */
    if ((ret = duo_call(ctx, "POST", DUO_API_VERSION "/auth.json",
                   flags & DUO_FLAG_SYNC ? DUO_NO_TIMEOUT : ctx->https_timeout)) != DUO_OK ||
         (ret = _duo_json_response(ctx)) != DUO_OK) {
        return (ret);
    }

    /* Handle sync status */
    if ((flags & DUO_FLAG_SYNC) != 0) {
        JSON_Value *json = json_parse_string(ctx->body);
        JSON_Object *json_obj = json_value_get_object(json);
        JSON_Object *json_response;
        _JSON_FIND_OBJECT(json_response, json_obj, "response", json);
        const char *status;
        _JSON_FIND_STRING(status, json_response, "status", json);
        if (ctx->conv_status != NULL) {
            ctx->conv_status(ctx->conv_arg,
            status);
        }
        const char* result;
        _JSON_FIND_STRING(result, json_response, "result", json);

        if (strcasecmp(result, "allow") == 0) {
            ret = DUO_OK;
        } else if (strcasecmp(result, "deny") == 0) {
            ret = DUO_FAIL;
        } else {
            _duo_seterr(ctx, "JSON invalid 'result': %s", result);
            ret = DUO_SERVER_ERROR;
        }
        _JSON_VALUE_FREE(json);
        return (ret);
    }
    /* Async status - long-poll on txid */
    JSON_Value *json = json_parse_string(ctx->body);
    JSON_Object *json_obj = json_value_get_object(json);
    JSON_Object *json_response;
    _JSON_FIND_OBJECT(json_response, json_obj, "response", json);

    const char* txid;
    _JSON_FIND_STRING(txid, json_response, "txid", json);
    if (strlcpy(buf, txid, sizeof(buf)) >= sizeof(buf)) {
        _JSON_VALUE_FREE(json);
        return (DUO_LIB_ERROR);
    }
    /* XXX newline between prompt and async status lines */
    if (ctx->conv_status != NULL) {
        ctx->conv_status(ctx->conv_arg, "");
    }
    ret = DUO_SERVER_ERROR;

    for (i = 0; i < 20; i++) {
        if ((ret = duo_add_param(ctx, "txid", buf)) != DUO_OK ||
            (ret = duo_call(ctx, "GET",
            DUO_API_VERSION "/status.json", DUO_NO_TIMEOUT)) != DUO_OK ||
            (ret = _duo_json_response(ctx)) != DUO_OK) {
            break;
        }

        JSON_Value *json_new = json_parse_string(ctx->body);
        JSON_Object *json_obj_new = json_value_get_object(json_new);
        JSON_Object *json_response_new;
        _JSON_FIND_OBJECT(json_response_new, json_obj_new, "response", json);
        const char *status_json_obj;
        _JSON_FIND_STRING(status_json_obj, json_response_new, "status", json);
        if (status_json_obj != NULL) {
            if (ctx->conv_status != NULL) {
                ctx->conv_status(ctx->conv_arg, status_json_obj);
            }
        }

        //We might not have 'result' defined but we don't want to quit the program
        //if it's not in our object yet
        const char* result = json_object_get_string(json_response_new, "result");
        if (result != NULL) {
            if (strcasecmp(result, "allow") == 0) {
                ret = DUO_OK;
            } else if (strcasecmp(result, "deny") == 0) {
                ret = DUO_FAIL;
            } else {
                _duo_seterr(ctx, "JSON invalid 'result': %s",
                    result);
                ret = DUO_SERVER_ERROR;
            }
            break;
        }
    }
    _JSON_VALUE_FREE(json);
    return (ret);
}
